pennylane_lightning/src/algorithms/Observables.hpp

```diff
@@ -0,0 +1,419 @@
+
+        return (obs_name_ == other_cast.obs_name_) &&
+               (wires_ == other_cast.wires_) && (params_ == other_cast.params_);
+    }
+
+  public:
+    /**
+     * @brief Construct a NamedObs object, representing a given observable.
+     *
+     * @param arg1 Name of the observable.
+     * @param arg2 Argument to construct wires.
+     * @param arg3 Argument to construct parameters
+     */
+    NamedObs(std::string obs_name, std::vector<size_t> wires,
+             std::vector<T> params = {})
+        : obs_name_{std::move(obs_name)}, wires_{std::move(wires)},
+          params_{std::move(params)} {
+        using Gates::Constant::gate_names;
+        using Gates::Constant::gate_num_params;
+        using Gates::Constant::gate_wires;
+
+        const auto gate_op = Util::lookup(Util::reverse_pairs(gate_names),
+                                          std::string_view{obs_name_});
+        PL_ASSERT(Util::lookup(gate_wires, gate_op) == wires_.size());
+        PL_ASSERT(Util::lookup(gate_num_params, gate_op) == params_.size());
+    }
+
+    [[nodiscard]] auto getObsName() const -> std::string override {
+        return obs_name_;
+    }
+
+    [[nodiscard]] auto getWires() const -> std::vector<size_t> final {
+        return wires_;
+    }
+
+    void applyInPlace(StateVectorManagedCPU<T> &sv) const final {
+        sv.applyOperation(obs_name_, wires_, false, params_);
+    }
+};
+
+/**
+ * @brief Class models
+ *
+ */
+template <typename T> class HermitianObs final : public Observable<T> {
+  public:
+    using MatrixT = std::vector<std::complex<T>>;
+
+  private:
+    MatrixT matrix_;
+    std::vector<size_t> wires_;
+
+    [[nodiscard]] bool isEqual(const Observable<T> &other) const final {
+        const auto &other_cast = static_cast<const HermitianObs<T> &>(other);
+
+        return (matrix_ == other_cast.matrix_) && (wires_ == other_cast.wires_);
+    }
+
+  public:
+    /**
+     * @brief Create Hermitian observable
+     *
+     * @param arg1 Matrix in row major format.
+     * @param arg2 Wires the observable applies to.
+     */
+    template <typename T1>
+    HermitianObs(T1 &&arg1, std::vector<size_t> wires)
+        : matrix_{std::forward<T1>(arg1)}, wires_{std::move(wires)} {
+        PL_ASSERT(matrix_.size() ==
+                  Util::exp2(wires_.size()) * Util::exp2(wires_.size()));
+    }
+
+    [[nodiscard]] auto getMatrix() const -> const MatrixT & { return matrix_; }
+
+    [[nodiscard]] auto getWires() const -> std::vector<size_t> final {
+        return wires_;
+    }
+
+    [[nodiscard]] auto getObsName() const -> std::string final {
+        return "Hermitian";
+    }
+
+    void applyInPlace(StateVectorManagedCPU<T> &sv) const final {
+        sv.applyMatrix(matrix_, wires_);
+    }
+};
+
+/**
+ * @brief Tensor product observable class
+ */
+template <typename T> class TensorProdObs final : public Observable<T> {
+  private:
+    std::vector<std::shared_ptr<Observable<T>>> obs_;
```

Comment:
Ownership of the tensor is maintained at the Python layer?
We are guaranteed the `shared_ptr` is a `shared_ptr` all the way through?

Reply:
We explicitly expose `std::shared_ptr` of observable classes with pybind11 (https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#std-shared-ptr).